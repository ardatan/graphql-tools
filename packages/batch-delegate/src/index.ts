export * from './batchDelegateToSchema.js';
export * from './createBatchDelegateFn.js';

export * from './types.js';
