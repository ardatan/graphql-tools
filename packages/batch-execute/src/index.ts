export * from './createBatchingExecutor.js';
export * from './getBatchingExecutor.js';
