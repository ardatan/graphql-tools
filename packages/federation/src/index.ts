export * from './gateway.js';
export * from './subgraph.js';
export * from './supergraph.js';
export * from './utils.js';
