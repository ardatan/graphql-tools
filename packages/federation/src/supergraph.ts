import {
  buildASTSchema,
  DocumentNode,
  EnumTypeDefinitionNode,
  EnumValueDefinitionNode,
  FieldDefinitionNode,
  Kind,
  NamedTypeNode,
  ObjectTypeDefinitionNode,
  parse,
  ScalarTypeDefinitionNode,
  TypeDefinitionNode,
  UnionTypeDefinitionNode,
  visit,
} from 'graphql';
import { MergedTypeConfig, SubschemaConfig } from '@graphql-tools/delegate';
import { buildHTTPExecutor } from '@graphql-tools/executor-http';
import { stitchSchemas } from '@graphql-tools/stitch';
import type { Executor } from '@graphql-tools/utils';
import {
  filterInternalFieldsAndTypes,
  getArgsFromKeysForFederation,
  getKeyForFederation,
  getNamedTypeNode,
} from './utils.js';

export interface GetSubschemasFromSupergraphSdlOpts {
  supergraphSdl: string | DocumentNode;
  onExecutor?: (opts: { subgraphName: string; endpoint: string }) => Executor;
  batch?: boolean;
}

export function getSubschemasFromSupergraphSdl({
  supergraphSdl,
  onExecutor = ({ endpoint }) => buildHTTPExecutor({ endpoint }),
  batch = false,
}: GetSubschemasFromSupergraphSdlOpts) {
  const ast =
    typeof supergraphSdl === 'string' ? parse(supergraphSdl, { noLocation: true }) : supergraphSdl;
  const subgraphQueryFieldDefinitionNodes = new Map<string, FieldDefinitionNode[]>();
  const subgraphEndpointMap = new Map<string, string>();
  const subgraphTypesMap = new Map<string, TypeDefinitionNode[]>();
  const typeNameKeyBySubgraphMap = new Map<string, Map<string, string>>();
  const typeNameFieldsKeyBySubgraphMap = new Map<string, Map<string, Map<string, string>>>();
  const typeNameCanonicalMap = new Map<string, string>();
  const unownedTypes = new Map<string, TypeDefinitionNode>();
  const dependenciesBySubgraphMap = new Map<string, Set<string>>();
  visit(ast, {
    ScalarTypeDefinition(node) {
      unownedTypes.set(node.name.value, node);
    },
    InterfaceTypeDefinition(node) {
      unownedTypes.set(node.name.value, node);
    },
    UnionTypeDefinition(node) {
      unownedTypes.set(node.name.value, node);
    },
    EnumTypeDefinition(node) {
      if (node.name.value === 'join__Graph') {
        node.values?.forEach(valueNode => {
          const joinGraphDirectiveNode = valueNode.directives?.find(
            directiveNode => directiveNode.name.value === 'join__graph',
          );
          if (joinGraphDirectiveNode) {
            const urlArgumentNode = joinGraphDirectiveNode.arguments?.find(
              argumentNode => argumentNode.name.value === 'url',
            );
            if (urlArgumentNode?.value?.kind === Kind.STRING) {
              subgraphEndpointMap.set(valueNode.name.value, urlArgumentNode.value.value);
            }
          }
        });
      }
      node.directives?.forEach(directiveNode => {
        if (directiveNode.name.value === 'join__type') {
          const graphArgumentNode = directiveNode.arguments?.find(
            argumentNode => argumentNode.name.value === 'graph',
          );
          if (graphArgumentNode?.value?.kind === Kind.ENUM) {
            const graphName = graphArgumentNode.value.value;
            const enumValueNodes: EnumValueDefinitionNode[] =
              node.values?.filter(valueNode => {
                const joinEnumValueDirectiveNode = valueNode.directives?.find(
                  directiveNode => directiveNode.name.value === 'join__enumValue',
                );
                if (joinEnumValueDirectiveNode) {
                  const graphArgumentNode = joinEnumValueDirectiveNode.arguments?.find(
                    argumentNode => argumentNode.name.value === 'graph',
                  );
                  if (graphArgumentNode?.value?.kind === Kind.ENUM) {
                    return graphArgumentNode.value.value === graphName;
                  }
                }
                return false;
              }) || [];
            const enumTypedDefNodeForSubgraph: EnumTypeDefinitionNode = {
              ...node,
              values: enumValueNodes,
            };
            let subgraphTypes = subgraphTypesMap.get(graphName);
            if (!subgraphTypes) {
              subgraphTypes = [];
              subgraphTypesMap.set(graphName, subgraphTypes);
            }
            subgraphTypes.push(enumTypedDefNodeForSubgraph);
          }
        }
      });
    },
    ObjectTypeDefinition(node) {
      if (node.name.value === 'Query') {
        node.fields?.forEach(fieldNode => {
          fieldNode.directives?.forEach(directiveNode => {
            if (directiveNode.name.value === 'join__field') {
              const graphArgumentNode = directiveNode.arguments?.find(
                argumentNode => argumentNode.name.value === 'graph',
              );
              if (graphArgumentNode?.value?.kind === Kind.ENUM) {
                const graphName = graphArgumentNode.value.value;
                let fieldDefinitionNodesOfSubgraph =
                  subgraphQueryFieldDefinitionNodes.get(graphName);
                if (!fieldDefinitionNodesOfSubgraph) {
                  fieldDefinitionNodesOfSubgraph = [];
                  subgraphQueryFieldDefinitionNodes.set(graphName, fieldDefinitionNodesOfSubgraph);
                }
                fieldDefinitionNodesOfSubgraph.push(fieldNode);
              }
            }
          });
        });
      }
      let owned = false;
      node.directives?.forEach(directiveNode => {
        if (directiveNode.name.value === 'join__owner') {
          const graphArgumentNode = directiveNode.arguments?.find(
            argumentNode => argumentNode.name.value === 'graph',
          );
          if (graphArgumentNode?.value?.kind === Kind.ENUM) {
            typeNameCanonicalMap.set(node.name.value, graphArgumentNode.value.value);
          }
        }
        if (directiveNode.name.value === 'join__type') {
          const graphArgumentNode = directiveNode.arguments?.find(
            argumentNode => argumentNode.name.value === 'graph',
          );
          if (graphArgumentNode?.value?.kind === Kind.ENUM) {
            let keyFieldName: string;
            const graphName = graphArgumentNode.value.value;
            const keyArgumentNode = directiveNode.arguments?.find(
              argumentNode => argumentNode.name.value === 'key',
            );
            if (keyArgumentNode?.value?.kind === Kind.STRING) {
              let typeNameKeyMap = typeNameKeyBySubgraphMap.get(graphName);
              if (!typeNameKeyMap) {
                typeNameKeyMap = new Map();
                typeNameKeyBySubgraphMap.set(graphName, typeNameKeyMap);
              }
              keyFieldName = keyArgumentNode.value.value;
              typeNameKeyMap.set(node.name.value, keyFieldName);
            }
            const fieldDefinitionNodesOfSubgraph =
              node.fields?.filter(fieldNode => {
                if (fieldNode.name.value === keyFieldName) {
                  return true;
                }
                const joinFieldDirectiveNode = fieldNode.directives?.find(
                  directiveNode => directiveNode.name.value === 'join__field',
                );
                if (joinFieldDirectiveNode) {
                  const graphArgumentNode = joinFieldDirectiveNode.arguments?.find(
                    argumentNode => argumentNode.name.value === 'graph',
                  );
                  if (graphArgumentNode?.value?.kind === Kind.ENUM) {
                    if (graphArgumentNode.value.value === graphName) {
                      let dependenciesBySubgraph = dependenciesBySubgraphMap.get(graphName);
                      if (!dependenciesBySubgraph) {
                        dependenciesBySubgraph = new Set();
                        dependenciesBySubgraphMap.set(graphName, dependenciesBySubgraph);
                      }
                      const namedTypeNode = getNamedTypeNode(fieldNode.type);
                      dependenciesBySubgraph.add(namedTypeNode.name.value);

                      const requiresArgumentNode = joinFieldDirectiveNode.arguments?.find(
                        argumentNode => argumentNode.name.value === 'requires',
                      );
                      if (requiresArgumentNode?.value?.kind === Kind.STRING) {
                        let typeNameFieldsKeyMap = typeNameFieldsKeyBySubgraphMap.get(graphName);
                        if (!typeNameFieldsKeyMap) {
                          typeNameFieldsKeyMap = new Map();
                          typeNameFieldsKeyBySubgraphMap.set(graphName, typeNameFieldsKeyMap);
                        }
                        let fieldsKeyMap = typeNameFieldsKeyMap.get(node.name.value);
                        if (!fieldsKeyMap) {
                          fieldsKeyMap = new Map();
                          typeNameFieldsKeyMap.set(node.name.value, fieldsKeyMap);
                        }
                        fieldsKeyMap.set(fieldNode.name.value, requiresArgumentNode.value.value);
                      }
                      return true;
                    }
                  }
                } else {
                  return true;
                }
                return false;
              }) || [];
            const objectTypedDefNodeForSubgraph: ObjectTypeDefinitionNode = {
              ...node,
              fields: fieldDefinitionNodesOfSubgraph,
            };
            let subgraphTypes = subgraphTypesMap.get(graphName);
            if (!subgraphTypes) {
              subgraphTypes = [];
              subgraphTypesMap.set(graphName, subgraphTypes);
            }
            owned = true;
            subgraphTypes.push(objectTypedDefNodeForSubgraph);
          }
        }
      });
      if (!owned) {
        unownedTypes.set(node.name.value, node);
      }
    },
  });
  const subschemaMap = new Map<string, SubschemaConfig>();
  for (const [subgraphName, endpoint] of subgraphEndpointMap) {
    const executor = onExecutor({ subgraphName, endpoint });
    const mergeConfig: SubschemaConfig['merge'] = {};
    const typeNameKeyMap = typeNameKeyBySubgraphMap.get(subgraphName);
    const unionTypeNodes: NamedTypeNode[] = [];
    if (typeNameKeyMap) {
      const typeNameFieldsKeyMap = typeNameFieldsKeyBySubgraphMap.get(subgraphName);
      for (const [typeName, key] of typeNameKeyMap) {
        const mergedTypeConfig: MergedTypeConfig = (mergeConfig[typeName] = {
          selectionSet: `{ ${key} }`,
          argsFromKeys: getArgsFromKeysForFederation,
          key: getKeyForFederation,
          fieldName: `_entities`,
        });
        const keyProps = key.split(' ');
        mergedTypeConfig.dataLoaderOptions = {
          cacheKeyFn(root) {
            return keyProps.map(key => root[key]).join(' ');
          },
        };
        const fieldsKeyMap = typeNameFieldsKeyMap?.get(typeName);
        if (fieldsKeyMap) {
          const fieldsConfig = (mergedTypeConfig.fields = {});
          for (const [fieldName, key] of fieldsKeyMap) {
            fieldsConfig[fieldName] = {
              selectionSet: `{ ${key} }`,
              computed: true,
            };
          }
        }
        if (typeNameCanonicalMap.get(typeName) === subgraphName) {
          mergedTypeConfig.canonical = true;
        }
        unionTypeNodes.push({
          kind: Kind.NAMED_TYPE,
          name: {
            kind: Kind.NAME,
            value: typeName,
          },
        });
      }
    }
    const entitiesUnionTypeDefinitionNode: UnionTypeDefinitionNode = {
      name: {
        kind: Kind.NAME,
        value: '_Entity',
      },
      kind: Kind.UNION_TYPE_DEFINITION,
      types: unionTypeNodes,
    };
    const queryFields = subgraphQueryFieldDefinitionNodes.get(subgraphName) || [];
    const queryWithEntitiesFieldDefinitionNode: ObjectTypeDefinitionNode = {
      name: {
        kind: Kind.NAME,
        value: 'Query',
      },
      kind: Kind.OBJECT_TYPE_DEFINITION,
      fields: [
        ...queryFields,
        {
          kind: Kind.FIELD_DEFINITION,
          name: {
            kind: Kind.NAME,
            value: '_entities',
          },
          type: {
            kind: Kind.NAMED_TYPE,
            name: {
              kind: Kind.NAME,
              value: '_Entity',
            },
          },
          arguments: [
            {
              kind: Kind.INPUT_VALUE_DEFINITION,
              name: {
                kind: Kind.NAME,
                value: 'representations',
              },
              type: {
                kind: Kind.NON_NULL_TYPE,
                type: {
                  kind: Kind.LIST_TYPE,
                  type: {
                    kind: Kind.NON_NULL_TYPE,
                    type: {
                      kind: Kind.NAMED_TYPE,
                      name: {
                        kind: Kind.NAME,
                        value: '_Any',
                      },
                    },
                  },
                },
              },
            },
          ],
        },
      ],
    };
    const subgraphTypes = subgraphTypesMap.get(subgraphName) || [];
    const dependencyNames = dependenciesBySubgraphMap.get(subgraphName) || new Set();
    for (const dependencyName of dependencyNames) {
      const dependencyType = unownedTypes.get(dependencyName);
      if (dependencyType) {
        subgraphTypes.push(dependencyType);
      }
    }
    const schema = buildASTSchema(
      {
        kind: Kind.DOCUMENT,
        definitions: [
          ...subgraphTypes,
          entitiesUnionTypeDefinitionNode,
          anyTypeDefinitionNode,
          queryWithEntitiesFieldDefinitionNode,
        ],
      },
      {
        assumeValidSDL: true,
        assumeValid: true,
      },
    );
    subschemaMap.set(subgraphName, {
      schema,
      executor,
      merge: mergeConfig,
      batch,
    });
  }
  return subschemaMap;
}

export function getStitchedSchemaFromSupergraphSdl(opts: GetSubschemasFromSupergraphSdlOpts) {
  const subschemaMap = getSubschemasFromSupergraphSdl(opts);
  const supergraphSchema = stitchSchemas({
    subschemas: [...subschemaMap.values()],
  });
  return filterInternalFieldsAndTypes(supergraphSchema);
}

const anyTypeDefinitionNode: ScalarTypeDefinitionNode = {
  name: {
    kind: Kind.NAME,
    value: '_Any',
  },
  kind: Kind.SCALAR_TYPE_DEFINITION,
};
