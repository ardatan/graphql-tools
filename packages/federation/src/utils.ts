import { GraphQLSchema } from 'graphql';
import { FilterRootFields, FilterTypes } from '@graphql-tools/wrap';

export function getArgsFromKeysForFederation(representations: readonly any[]) {
  return { representations };
}

export function getKeyForFederation(root: any) {
  return root;
}

export function filterInternalFieldsAndTypes(finalSchema: GraphQLSchema) {
  const removeEntitiesField = new FilterRootFields(
    (operation, fieldName) =>
      !(operation === 'Query' && (fieldName === '_entities' || fieldName === '_sdl')),
  );
  const removeEntityAndAny = new FilterTypes(
    type =>
      type.name !== '_Entity' &&
      type.name !== '_Any' &&
      type.name !== '_FieldSet' &&
      type.name !== '_Service' &&
      !type.name.startsWith('link__'),
  );
  const fakeSubschemaConfig = {
    schema: finalSchema,
    transforms: [removeEntitiesField, removeEntityAndAny],
  };
  for (const transform of fakeSubschemaConfig.transforms) {
    finalSchema = transform.transformSchema(finalSchema, fakeSubschemaConfig);
  }
  return finalSchema;
}
