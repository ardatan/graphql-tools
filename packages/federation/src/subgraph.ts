import { GraphQLResolveInfo, visit } from 'graphql';
import { ValueOrPromise } from 'value-or-promise';
import { mergeResolvers, mergeTypeDefs } from '@graphql-tools/merge';
import { IExecutableSchemaDefinition, makeExecutableSchema } from '@graphql-tools/schema';
import { printSchemaWithDirectives } from '@graphql-tools/utils';

export const SubgraphBaseSDL = /* GraphQL */ `
  scalar _Any
  scalar _FieldSet
  scalar link__Import

  enum link__Purpose {
    SECURITY
    EXECUTION
  }

  type _Service {
    sdl: String!
  }

  type Query {
    _entities(representations: [_Any!]!): [_Entity]!
    _service: _Service!
  }

  directive @external on FIELD_DEFINITION | OBJECT
  directive @requires(fields: _FieldSet!) on FIELD_DEFINITION
  directive @provides(fields: _FieldSet!) on FIELD_DEFINITION
  directive @key(fields: _FieldSet!, resolvable: Boolean = true) repeatable on OBJECT | INTERFACE
  directive @link(
    url: String!
    as: String
    for: link__Purpose
    import: [link__Import]
  ) repeatable on SCHEMA
  directive @shareable repeatable on OBJECT | FIELD_DEFINITION
  directive @inaccessible on FIELD_DEFINITION | OBJECT | INTERFACE | UNION | ARGUMENT_DEFINITION | SCALAR | ENUM | ENUM_VALUE | INPUT_OBJECT | INPUT_FIELD_DEFINITION
  directive @tag(
    name: String!
  ) repeatable on FIELD_DEFINITION | INTERFACE | OBJECT | UNION | ARGUMENT_DEFINITION | SCALAR | ENUM | ENUM_VALUE | INPUT_OBJECT | INPUT_FIELD_DEFINITION
  directive @override(from: String!) on FIELD_DEFINITION
  directive @composeDirective(name: String!) repeatable on SCHEMA

  directive @extends on OBJECT | INTERFACE
`;

export function buildSubgraphSchema<TContext = any>(opts: IExecutableSchemaDefinition<TContext>) {
  const typeDefs = mergeTypeDefs([SubgraphBaseSDL, opts.typeDefs]);
  const entityTypeNames: string[] = [];
  visit(typeDefs, {
    ObjectTypeDefinition: node => {
      if (node.directives?.some(directive => directive.name.value === 'key')) {
        entityTypeNames.push(node.name.value);
      }
    },
  });
  const entityTypeDefinition = `union _Entity = ${entityTypeNames.join(' | ')}`;
  const givenResolvers: any = mergeResolvers(opts.resolvers);
  const subgraphResolvers = {
    _Entity: {
      __resolveType: (obj: { __typename: string }) => obj.__typename,
    },
    Query: {
      _entities: (
        _root: never,
        args: { representations: any[] },
        context: TContext,
        info: GraphQLResolveInfo,
      ) =>
        ValueOrPromise.all(
          args.representations.map(representation =>
            new ValueOrPromise(
              () =>
                givenResolvers[representation.__typename]?.__resolveReference?.(
                  representation,
                  context,
                  info,
                ),
            ).then(resolvedEntity => {
              if (!resolvedEntity) {
                return representation;
              }
              if (!resolvedEntity.__typename) {
                resolvedEntity.__typename = representation.__typename;
              }
              return resolvedEntity;
            }),
          ),
        ).resolve(),
      _service: () => ({}),
    },
    _Service: {
      sdl: (_root: never, _args: never, _context: TContext, info: GraphQLResolveInfo) =>
        printSchemaWithDirectives(info.schema),
    },
  };
  return makeExecutableSchema({
    assumeValid: true,
    assumeValidSDL: true,
    ...opts,
    typeDefs: [entityTypeDefinition, typeDefs],
    resolvers: [subgraphResolvers, givenResolvers],
  });
}
