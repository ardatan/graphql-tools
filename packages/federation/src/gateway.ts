import {
  buildASTSchema,
  concatAST,
  DocumentNode,
  FieldDefinitionNode,
  GraphQLSchema,
  Kind,
  ObjectTypeDefinitionNode,
  ObjectTypeExtensionNode,
  parse,
  visit,
} from 'graphql';
import { createDefaultExecutor, SubschemaConfig } from '@graphql-tools/delegate';
import { buildHTTPExecutor, HTTPExecutorOptions } from '@graphql-tools/executor-http';
import { stitchSchemas, SubschemaConfigTransform } from '@graphql-tools/stitch';
import {
  createGraphQLError,
  ExecutionResult,
  Executor,
  getDocumentNodeFromSchema,
  inspect,
} from '@graphql-tools/utils';
import { SubgraphBaseSDL } from './subgraph.js';
import {
  filterInternalFieldsAndTypes,
  getArgsFromKeysForFederation,
  getCacheKeyFnFromKey,
  getKeyForFederation,
} from './utils.js';

export const SubgraphSDLQuery = /* GraphQL */ `
  query SubgraphSDL {
    _service {
      sdl
    }
  }
`;

export async function getSubschemaForFederationWithURL(
  config: HTTPExecutorOptions,
): Promise<SubschemaConfig> {
  const executor = buildHTTPExecutor(config);
  const subschemaConfig = await getSubschemaForFederationWithExecutor(executor);
  return {
    batch: true,
    ...subschemaConfig,
  };
}

export function getSubschemaForFederationWithTypeDefs(typeDefs: DocumentNode): SubschemaConfig {
  const subschemaConfig = {} as SubschemaConfig;
  const typeMergingConfig = (subschemaConfig.merge = subschemaConfig.merge || {});
  const entityTypes: string[] = [];
  const visitor = (node: ObjectTypeDefinitionNode | ObjectTypeExtensionNode) => {
    if (node.directives) {
      const typeName = node.name.value;
      const selections: string[] = [];
      for (const directive of node.directives) {
        const directiveArgs = directive.arguments || [];
        switch (directive.name.value) {
          case 'key': {
            if (
              directiveArgs.some(
                arg =>
                  arg.name.value === 'resolvable' &&
                  arg.value.kind === Kind.BOOLEAN &&
                  arg.value.value === false,
              )
            ) {
              continue;
            }
            const selectionValueNode = directiveArgs.find(
              arg => arg.name.value === 'fields',
            )?.value;
            if (selectionValueNode?.kind === Kind.STRING) {
              selections.push(selectionValueNode.value);
            }
            break;
          }
          case 'inaccessible':
            return null;
        }
      }
      // If it is not an entity, continue
      if (selections.length === 0) {
        return node;
      }
      const typeMergingTypeConfig = (typeMergingConfig[typeName] =
        typeMergingConfig[typeName] || {});
      if (
        node.kind === Kind.OBJECT_TYPE_DEFINITION &&
        !node.directives?.some(d => d.name.value === 'extends')
      ) {
        typeMergingTypeConfig.canonical = true;
      }
      entityTypes.push(typeName);
      const selectionsStr = selections.join(' ');
      typeMergingTypeConfig.selectionSet = `{ ${selectionsStr} }`;
      typeMergingTypeConfig.dataLoaderOptions = {
        cacheKeyFn: getCacheKeyFnFromKey(selectionsStr),
      };
      typeMergingTypeConfig.argsFromKeys = getArgsFromKeysForFederation;
      typeMergingTypeConfig.fieldName = `_entities`;
      typeMergingTypeConfig.key = getKeyForFederation;
      const fields = [];
      if (node.fields) {
        for (const fieldNode of node.fields) {
          let removed = false;
          if (fieldNode.directives) {
            const fieldName = fieldNode.name.value;
            for (const directive of fieldNode.directives) {
              const directiveArgs = directive.arguments || [];
              switch (directive.name.value) {
                case 'requires': {
                  const typeMergingFieldsConfig = (typeMergingTypeConfig.fields =
                    typeMergingTypeConfig.fields || {});
                  typeMergingFieldsConfig[fieldName] = typeMergingFieldsConfig[fieldName] || {};
                  if (
                    directiveArgs.some(
                      arg =>
                        arg.name.value === 'resolvable' &&
                        arg.value.kind === Kind.BOOLEAN &&
                        arg.value.value === false,
                    )
                  ) {
                    continue;
                  }
                  const selectionValueNode = directiveArgs.find(
                    arg => arg.name.value === 'fields',
                  )?.value;
                  if (selectionValueNode?.kind === Kind.STRING) {
                    typeMergingFieldsConfig[fieldName].selectionSet =
                      `{ ${selectionValueNode.value} }`;
                    typeMergingFieldsConfig[fieldName].computed = true;
                  }
                  break;
                }
                case 'external':
                case 'inaccessible': {
                  removed = !typeMergingTypeConfig.selectionSet?.includes(` ${fieldName} `);
                  break;
                }
                case 'override': {
                  const typeMergingFieldsConfig = (typeMergingTypeConfig.fields =
                    typeMergingTypeConfig.fields || {});
                  typeMergingFieldsConfig[fieldName] = typeMergingFieldsConfig[fieldName] || {};
                  typeMergingFieldsConfig[fieldName].canonical = true;
                  break;
                }
              }
            }
          }
          if (!removed) {
            fields.push(fieldNode);
          }
        }
        (node.fields as FieldDefinitionNode[]) = fields;
      }
    }
    return {
      ...node,
      kind: Kind.OBJECT_TYPE_DEFINITION,
    };
  };
  const parsedSDL = visit(typeDefs, {
    ObjectTypeExtension: visitor,
    ObjectTypeDefinition: visitor,
  });
  let extraSdl = SubgraphBaseSDL;
  if (entityTypes.length > 0) {
    extraSdl += `\nunion _Entity = ${entityTypes.join(' | ')}`;
    extraSdl += `\nextend type Query { _entities(representations: [_Any!]!): [_Entity]! }`;
  }
  subschemaConfig.schema = buildASTSchema(concatAST([parse(extraSdl), parsedSDL]), {
    assumeValidSDL: true,
    assumeValid: true,
  });
  // subschemaConfig.batch = true;
  return subschemaConfig;
}

export async function getSubschemaForFederationWithExecutor(executor: Executor) {
  const sdlQueryResult = (await executor({
    document: parse(SubgraphSDLQuery),
  })) as ExecutionResult;

  if (sdlQueryResult.errors?.length) {
    const error = sdlQueryResult.errors[0];
    throw createGraphQLError(error.message, error);
  }

  if (!sdlQueryResult.data?._service?.sdl) {
    throw new Error(`Unexpected result: ${inspect(sdlQueryResult)}`);
  }

  const typeDefs = parse(sdlQueryResult.data._service.sdl);

  const subschemaConfig = getSubschemaForFederationWithTypeDefs(typeDefs);

  return {
    ...subschemaConfig,
    executor,
  };
}

export async function getSubschemaForFederationWithSchema(schema: GraphQLSchema) {
  const executor = createDefaultExecutor(schema);
  return getSubschemaForFederationWithExecutor(executor);
}

export async function getStitchedSchemaWithUrls(configs: HTTPExecutorOptions[]) {
  const subschemas = await Promise.all(
    configs.map(config => getSubschemaForFederationWithURL(config)),
  );
  const schema = stitchSchemas({
    subschemas,
  });
  return filterInternalFieldsAndTypes(schema);
}

export const federationSubschemaTransformer: SubschemaConfigTransform =
  function federationSubschemaTransformer(subschemaConfig: SubschemaConfig): SubschemaConfig {
    const typeDefs = getDocumentNodeFromSchema(subschemaConfig.schema);
    const newSubschema = getSubschemaForFederationWithTypeDefs(typeDefs);
    return {
      ...subschemaConfig,
      ...newSubschema,
      merge: {
        ...newSubschema.merge,
        ...subschemaConfig.merge,
      },
    };
  };
