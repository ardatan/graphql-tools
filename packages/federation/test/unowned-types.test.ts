import { promises as fs } from 'fs';
import { join } from 'path';
import { printSchemaWithDirectives } from '@graphql-tools/utils';
import { getStitchedSchemaFromSupergraphSdl } from '../src/supergraph';

describe('Unowned types', () => {
  it('works', async () => {
    const supergraphSdl = await fs.readFile(
      join(__dirname, 'fixtures/unowned-types/supergraph.graphql'),
      'utf-8',
    );
    const stitchedSchema = getStitchedSchemaFromSupergraphSdl({
      supergraphSdl,
    });
    const printedSchema = printSchemaWithDirectives(stitchedSchema);
    expect(printedSchema.trim()).toMatchSnapshot();
  });
});
