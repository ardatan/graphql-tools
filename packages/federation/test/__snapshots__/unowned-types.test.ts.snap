// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unowned types works 1`] = `
"schema {
  query: Query
}

type Query {
  me: User
  users: [User]
  topProducts(first: Int): [Product]
}

type User {
  id: ID!
  numberOfReviews: Int
  reviews: [Review]
  birthDate: String
  name: String
  username: String
}

type Product {
  reviews(input: ProductReviewsInput): [Review]
  upc: String!
  inStock: Boolean
  shippingEstimate: Int
  name: String
  price: Int
  weight: Int
}

type AnonymousReview implements Review {
  body: String
  id: ID!
  product: Product
}

type UserReview implements Review {
  author: User
  body: String
  id: ID!
  product: Product
  stars: Stars
}

interface Review {
  body: String
  id: ID!
  product: Product
}

input ProductReviewsInput {
  first: Int
  order: ReviewsOrder
}

enum ReviewsOrder {
  ASC
  DESC
}

scalar Stars"
`;
